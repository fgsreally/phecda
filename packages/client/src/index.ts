export * from './base'
export * from './chain'
export * from './types'
