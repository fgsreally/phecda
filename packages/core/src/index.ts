export * from './decorators'
export * from './types'
export * from './helper'
export * from './utils'
export * from './core'
export * from './namespace'
export * from './custom/decorator'
export * from './preset'
