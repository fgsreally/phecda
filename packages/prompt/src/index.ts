/* eslint-disable no-new-func */
import prompt from 'prompt'
import { replacePromptStr } from './replacer'
const EXPRESS_RE = /^{{(.*)}}$/
export function fromJsonToData(jsonData: { [key: string]: any }, metadata?: any) {
  for (const i in jsonData) {
    if (typeof jsonData[i] === 'string' && EXPRESS_RE.test(jsonData[i])) {
      const expression = replacePromptStr(jsonData[i].match(EXPRESS_RE)[1])
      const fn = new Function('v', 'prompt', 'meta', `console.log(meta);return ${expression}`)
      jsonData[i] = (v: any) => fn(v, prompt, metadata)
    }
    if (typeof jsonData[i] === 'object')
      fromJsonToData(jsonData[i], metadata)
  }
}

/**
 * data is not immutable in `createPrompt`
 *
 * @param data JSON obj
 * @param metadata any data inject to prompt
 * @returns void
 */
export async function createPrompt(data: { [key: string]: any }, metadata?: any) {
  prompt.start()
  fromJsonToData(data, metadata)
  return prompt.get({ properties: data })
}

export { prompt }

// example:
// const data = {
//   proxy: {
//     description: 'Proxy url',
//   },
//   proxyCredentials: {
//     description: 'Proxy credentials',
//     ask: '{{<<proxy>>.value>0&&meta.value>10}}',
//   },
// }
