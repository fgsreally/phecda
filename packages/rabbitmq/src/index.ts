export * from 'phecda-server'
export * from './filter'
export * from './bind'
export * from './context'
export * from './types'
export * from './publish'
export * from './decorator'
