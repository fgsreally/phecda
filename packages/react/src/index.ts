export * from './core'
export * from 'phecda-web'
export * from './hook'
