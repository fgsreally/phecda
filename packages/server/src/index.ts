import 'reflect-metadata'

type Construct<T = any> = new (...args: Array<any>) => T

export function Factory<T>(Module: Construct<T>) {
  const moduleMap = new Map<Construct, InstanceType<Construct>>()
  const app = buildNestModule(Module, moduleMap) as InstanceType<Construct<T>>
  return { app, moduleMap }
}

function buildNestModule(Module: Construct, map: Map<Construct, InstanceType<Construct>>) {
  const paramtypes = getParamtypes(Module) as Construct[]
  let instance: InstanceType<Construct>
  if (map.has(Module)) {
    instance = map.get(Module)
    if (!instance)
      throw new Error(`exist Circular Module dep--${Module}`)
    return instance
  }
  map.set(Module, undefined)
  if (paramtypes) {
    instance = new Module(...paramtypes.map(item =>
      buildNestModule(item, map),
    ))
  }
  else {
    instance = new Module()
  }
  map.set(Module, instance)

  return instance
}

function getParamtypes(Module: Construct) {
  return Reflect.getMetadata('design:paramtypes', Module)
}
