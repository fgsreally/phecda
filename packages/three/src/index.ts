export * from './core'
export * from 'phecda-core'
