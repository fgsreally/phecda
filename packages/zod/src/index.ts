import type { ZodSchema, ZodTypeDef, z } from 'zod'
import { SHARE_KEY, addDecoToClass, setHandler, setStateKey } from 'phecda-core'

function ZodTo(cb: ((instance: any, addError:((msg: string) => void)) => any)) {
  return (proto: any, key: PropertyKey) => {
    setStateKey(proto.prototype, key)
    setHandler(proto.prototype, key, {
      async pipe(instance: any, addError: (msg: string) => void) {
        const ret = cb(instance, addError)

        if (ret) {
          for (const key in instance) {
            if (key !== '_value')
              delete instance[key]
          }

          for (const key in ret)
            instance[key] = ret[key]
        }
      },
    })
  }
}
export function zodToClass<
    TOutput = any,
    TDef extends ZodTypeDef = ZodTypeDef,
    TInput = TOutput,
>(zod: ZodSchema<TOutput, TDef, TInput>): new (data?: Partial<z.infer<ZodSchema<TOutput, TDef, TInput>>>) => z.infer<ZodSchema<TOutput, TDef, TInput>> {
  class Z {
    constructor(data: any) {
      for (const key in data)
        // @ts-expect-error trick
        this[key] = data[key]
    }

    _value = zod
  }

  addDecoToClass(Z, SHARE_KEY, ZodTo((ins, addError) => {
    const result = zod.safeParse(ins)
    if (!result.success)
      result.error.issues.forEach(({ message }) => addError(message))

    else return result.data
  }))

  return Z as any
}
