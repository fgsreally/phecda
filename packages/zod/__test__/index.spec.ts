import { describe, expect, it } from 'vitest'
import z from 'zod'
import { transformInstance } from 'phecda-core'
import { zodToClass } from '../src'

describe('zod in phecda', () => {
  it('basic validate', () => {
    const UserSchema = z.object({
      username: z.string().min(3).max(20), // username 必须为长度在 3 到 20 之间的字符串
      email: z.string().email(),
      isAdmin: z.boolean(), // isAdmin 必须为布尔值

    })

    const UserModel = zodToClass(UserSchema)
    const instance = new UserModel()
    instance.username = 'phecda'
    instance.email = 'phecda@a.com';
    // exclude
    (instance as any).xx = 'xx'
    expect(transformInstance(instance)).toMatchSnapshot()

    instance.isAdmin = true

    expect(transformInstance(instance).length).toBe(0)
    expect('xx' in instance).toBe(false)

    expect(UserModel.schema).toBeDefined()
    expect(instance).toMatchSnapshot()
  })
})
