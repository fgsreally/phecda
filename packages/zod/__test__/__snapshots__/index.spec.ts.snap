// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zod in phecda > basic validate 1`] = `
[
  "Required",
]
`;

exports[`zod in phecda > basic validate 2`] = `
Z {
  "email": "phecda@a.com",
  "isAdmin": true,
  "username": "phecda",
}
`;
